### PR TITLE
fix allocation issue caused by ordering of MVC result executors for Task results

### LINKED ISSUE #40364: TaskResultExecutor is never used by ActionMethodExecutor
### Describe the bug

Due to the order of executors in `ActionMethodExecutor.Executors`,  a `Task` returning action is always executed through `AwaitableResultExecutor` instead of `TaskResultExecutor`. From the ordering of the types and comments it looks like this wasn't the original intention and is sub-optimal (and the type is essentially unused currently).

### Expected Behavior

`Task` returning actions are executed using `TaskResultExecutor`.

### PR DESCRIPTION
**Pull request by rafikiassumani-msft**
# Fix allocation issue caused by ordering of MVC result executors for Task results

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

This PR changes the order of the ActionMethod executors for Task result types and adds tests to ensure the right executor gets chosen based on different action result scenarios. 

## Description

This PR changes the order of the ActionMethod executors so that for the case of `Task` where the result type is known at compile time, we can invoke the `Execute` sync method rather than the `ExecuteAsync` to allow us to "await" directly the result value via type cast.  The `ExecuteAync` method returns a custom awaitable, which can incur extra head allocations. 

Fixes #40364

### CODE REVIEW

File: src/Mvc/Mvc.Core/src/Infrastructure/ActionMethodExecutor.cs [hunk 1]
```diff
@@ -19,8 +19,8 @@ internal abstract class ActionMethodExecutor
             new SyncObjectResultExecutor(),
 
             // Executors for async methods
-            new AwaitableResultExecutor(),
             new TaskResultExecutor(),
+            new AwaitableResultExecutor(),
             new TaskOfIActionResultExecutor(),
             new TaskOfActionResultExecutor(),
```
Comment (BrennanConroy) [inline src/Mvc/Mvc.Core/src/Infrastructure/ActionMethodExecutor.cs:25]: I think we should make sure every executor on this list is called by a test, for example I don't see an added check for "TaskOfActionResultExecutor", does this mean we are missing test coverage for it? Or is it shadowed by TaskOfIActionResultExecutor?

Comment (rafikiassumani-msft) [reply src/Mvc/Mvc.Core/src/Infrastructure/ActionMethodExecutor.cs:25]: Good catch @BrennanConroy , I think `TaskOfIActionResultExecutor` had a duplicate test. I adjusted to account for `TaskOfActionResultExecutor`

File: src/Mvc/Mvc.Core/test/Infrastructure/ActionMethodExecutorTest.cs [hunk 2]
```diff
@@ -170,17 +201,19 @@ public async Task ActionMethodExecutor_ExecutesActionsAsynchronouslyReturningAct
...
         var valueTask = actionMethodExecutor.Execute(mapper, objectMethodExecutor, controller, Array.Empty<object>());
 
         // Assert
         await valueTask;
-        Assert.IsType<StatusCodeResult>(valueTask.Result);
+        Assert.Equal("TaskOfActionResultExecutor", actionMethodExecutor.GetType().Name);
+        Assert.IsType<ViewResult>(valueTask.Result);
     }
 
+
```
Comment (BrennanConroy) [inline src/Mvc/Mvc.Core/test/Infrastructure/ActionMethodExecutorTest.cs:216]:
```suggestion
```

**Review by BrennanConroy — APPROVED** [review]

File: src/Mvc/Mvc.Core/test/Infrastructure/ActionMethodExecutorTest.cs [hunk 3]
```diff
@@ -21,6 +22,7 @@ public void ActionMethodExecutor_ExecutesVoidActions()
         var valueTask = actionMethodExecutor.Execute(mapper, objectMethodExecutor, controller, Array.Empty<object>());
 
         // Assert
+        Assert.Equal("VoidResultExecutor", actionMethodExecutor.GetType().Name);
```
Comment (davidfowl) [inline src/Mvc/Mvc.Core/test/Infrastructure/ActionMethodExecutorTest.cs:25]: These types are internal so a type check would work? i.e.

```C#
Assert.IsType<VoidResultExecutor>(actionMethodExecutor);
```

Comment (rafikiassumani-msft) [reply src/Mvc/Mvc.Core/test/Infrastructure/ActionMethodExecutorTest.cs:25]: @davidfowl All the executors are private classes. The only internal class is the `ActionMethodExecutor`.  I don't think there is enough value in making these classes internal just for testing purposes.

**Review by halter73 — APPROVED** [review]
We could make the private nested classes internal nested classes instead to allow for `IsType` checks in our tests, but I don't think that's very important.

**Pull request merged**